contracts/05-tamagotchi-auto/src/lib.rs

```diff
@@ -1,19 +1,207 @@
+                tamagotchi.owner = actor_id;
+                msg::reply(TmgEvent::Transferred(actor_id), 0).expect("Error in sending reply");
+            }
+        }
+        TmgAction::Approve(actor_id) => {
+            if tamagotchi.owner == caller {
+                tamagotchi.approved_account = Some(actor_id);
+                msg::reply(TmgEvent::Approved(actor_id), 0).expect("Error in sending reply");
+            }
+        }
+        TmgAction::RevokeApproval => {
+            if tamagotchi.owner == caller {
+                tamagotchi.approved_account = None;
+                msg::reply(TmgEvent::ApprovalRevoked, 0).expect("Error in sending reply");
+            }
+        }
+        TmgAction::SetFTokenContract(contract_id) => {
+            if tamagotchi.is_owner_or_approved(&caller) {
+                tamagotchi.ft_contract_id = contract_id;
+                msg::reply(TmgEvent::FTokenContractSet, 0).expect("Error in sending reply");
+            }
+        }
+        TmgAction::ApproveTokens { account, amount } => {
+            if !tamagotchi.is_owner_or_approved(&caller) {
+                return;
+            }
+            tamagotchi.approve_tokens(account, amount).await;
+        }
+        TmgAction::BuyAttribute {
+            store_id,
+            attribute_id,
+        } => {
+            if !tamagotchi.is_owner_or_approved(&caller) {
+                return;
+            }
+            tamagotchi.buy_attribute(store_id, attribute_id).await;
+        }
+        // TODO; 6️⃣ Add handling new actions
+        TmgAction::CheckState => {
+            let payload;
+
+            // this only check the state of the tamagotchi, does not change
+            // the state of the contract
+
+            let blocks_height = blocks_height();
+            if tamagotchi.updated_feed_value(blocks_height) == 1 {
+                // tamagotchi.send_message(TmgEvent::FeedMe);
+                payload = TmgEvent::FeedMe;
+            } else if tamagotchi.updated_play_value(blocks_height) == 1 {
+                //tamagotchi.send_message(TmgEvent::PlayWithMe);
+                payload = TmgEvent::PlayWithMe;
+            } else if tamagotchi.updated_sleep_value(blocks_height) == 1 {
+                // tamagotchi.send_message(TmgEvent::WantToSleep);
+                payload = TmgEvent::WantToSleep;
+            } else {
+                payload = TmgEvent::AllGood;
+            }
+
+            // If the address is not the same as that of the contract,
+            // only the payload obtained is forwarded
+            if exec::program_id() != caller {
+                msg::reply(payload, 0).expect("Error in reply");
+                return;
+            }
+
+            if tamagotchi.reservations.len() == 1 {
+                // If there is only one gas reserve left, it is used to
+                // notify the owner to make more gas reserves.
+                *can_send_delayed_message = false;
```

Comment:
But what if owner makes an reservation between this `message` and the execution of the delayed message?